keps/sig-scheduling/4815-dra-partitionable-devices/README.md

```diff
@@ -0,0 +1,2899 @@
+  small, such that all full devices (and their partitions) can be represented
+  in a concise way.
+
+### Non-Goals
+
+* Allow a user to allocate arbitrarily-sized partitions of a device. In other
+  words, only those partitions explicitly defined by a `ResourceSlice` will be
+  allocatable by a user. Arbitrary slicing is out of scope for this KEP.
+
+## Proposal
+
+On a high level, this proposal is based around two new concepts:
+
+1. Capacity pool, which is a construct for defining the capacities available for
+   sharing among partitions of a single physical device. Drivers/controllers will
+   typically create a capacity pool that specifies all the available capacities of
+   a physical device and then define the partitions available for DRA. Each
+   partition will specify how much capacity it will consume from the capacity
+   pool, which allows the scheduler to keep track of available capacity in
+   a pool and decide whether a partition can be allocated. A device will
+   separately define the capacities that are accessible to the user for the partition.
+
+1. Mixins, which allows devices and resource pools to reference a mixin
+   object that defines some properties and that when used, will extend the properties
+   they define explicitly on the device or resource pool. The objective is to reduce
+   the duplication and overall footprint of `ResourceSlices`.
+
+The implementation of these concepts requires several changes to the
+ResourceSlice API. It introduces two new fields on the `ResourceSliceSpec`.
+`CapacityPools` defines a list of `CapacityPools` while `Mixins` define the
+mixins.
+
+1. The `CapacityPools` field is a list of named `CapacityPool`s. Each
+   defines a set of capacities that is available for devices. This makes it possible
+   to define overlapping partitions of devices, while still making sure that no
+   device can be allocated if the necessary capacity is not available. A `CapacityPool`
+   can optionally reference one or more `CapacityPoolMixin`s using the `Includes` field
+   to extend the set of capacities it defines explicitly.
+
+1. The `Mixins` field is of type `ResourceSliceMixin`, which has fields for the three
+   types of mixins available in a ResourceSlice. These are `CapacityPool`,
+   `Device`, and `DeviceCapacityConsumption`.
+
+    1. The `CapacityPool` field defines a list of named `CapacityPoolMixin`s. These
+      define a set of capacities that can be used to extend the capacities explicitly
+      defined in a `CapacityPool`. This allows for reduced duplication if there are many
+      identical physical devices that must be represented as capacity pools. CapacityPoolMixins
+      can not be referenced directly by devices.
+
+    1. The `Device` field is a list of named `DeviceMixin`s. These define
+      attributes and capacities that can be used to extend what is defined
+      explicitly in a `CompositeDevice`. Mixins can not be allocated directly,
```

Comment:
This is the first time that `CompositeDevice` is mentioned in the revised version of this KEP.#

As such, let's reword this to the following:
```suggestion
      explicitly in a new device type called `CompositeDevice` (introduced in more detail below). Mixins cannot be allocated directly,
```

Reply:
Side note: for now in this KEP revision we assume that we keep `CompositeDevice`. That's correct, no need to change anything.

In https://docs.google.com/document/d/1sCfO4hJUUhZpzld-_wEZi85XqSclEVvjUfsEe2-duZs/edit?tab=t.0 I proposed some changes that would make it safe to flatten the `Device` struct. This is something that we can try to describe in the next KEP update (https://github.com/kubernetes/enhancements/pull/5069), or treat it as an implementation choice and retro-actively update the KEP after we have agreed on the implementation in https://github.com/kubernetes/kubernetes/pull/129970. Given the time constraints I am leaning towards the latter.